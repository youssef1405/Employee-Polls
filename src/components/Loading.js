import styled from 'styled-components';
import { FaSpinner } from 'react-icons/fa';

const Loading = ({ text }) => {
  return (
    <Wrapper>
      <h2>{text}</h2>
      <FaSpinner className='spinner' />
    </Wrapper>
  );
};

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: #4e4e4e;

  .spinner {
    font-size: 2rem;
  }
`;

export default Loading;
