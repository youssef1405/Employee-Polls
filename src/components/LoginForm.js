import Wrapper from '../assests/wrappers/LoginForm';
import logo from '../assests/images/main-img.svg';

const LoginForm = () => {
  return (
    <Wrapper>
      <h1>Employee Polls</h1>
      <img src={logo} alt='' />

      <h2>Log In</h2>
      <form>
        <div>
          <label htmlFor=''>User</label>
          <input type='text' placeholder='Username' />
        </div>
        <div>
          <label htmlFor=''>Password</label>
          <input type='password' placeholder='Password' />
        </div>
      </form>
    </Wrapper>
  );
};

export default LoginForm;
