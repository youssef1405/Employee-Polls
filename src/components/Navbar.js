import Wrapper from '../assests/wrappers/Navbar';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';

const Navbar = () => {
  const username = useSelector((state) => state.users.currentUser);
  return (
    <Wrapper>
      <ul>
        <div>
          <li>
            <Link to='/'>Home</Link>
          </li>
          <li>
            <Link to='/leaderboard'>Leaderboard</Link>
          </li>
          <li>
            <Link to='/new-poll'>New</Link>
          </li>
        </div>
        <div>
          <li>{username}</li>
          <li>Logout</li>
        </div>
      </ul>
    </Wrapper>
  );
};

export default Navbar;
