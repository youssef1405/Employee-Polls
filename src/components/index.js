import Poll from './Poll';

export { Poll };
