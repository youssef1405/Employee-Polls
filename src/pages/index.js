import Home from './Home';
import Leaderboard from './Leaderboard';
import NewPoll from './NewPoll';
import Login from './Login';
import SharedLayout from './SharedLayout';
import OpenedPoll from './OpenedPoll';
import Error from './Error';

export { Home, Leaderboard, NewPoll, Login, SharedLayout, OpenedPoll, Error };
