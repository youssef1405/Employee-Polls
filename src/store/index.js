import { configureStore } from '@reduxjs/toolkit';
import pollsReducer from '../features/polls/pollsSlice';
import usersReducer from '../features/users/usersSlice';
import {
  loadPolls,
  showPoll,
  savePoll,
  changePollStatus,
} from '../features/polls/pollsSlice';
import {
  getUsers,
  setCurrentUser,
  addAnswer,
  addPoll,
} from '../features/users/usersSlice';

export const store = configureStore({
  reducer: {
    polls: pollsReducer,
    users: usersReducer,
  },
});

export {
  loadPolls,
  getUsers,
  setCurrentUser,
  showPoll,
  savePoll,
  changePollStatus,
  addAnswer,
  addPoll,
};
