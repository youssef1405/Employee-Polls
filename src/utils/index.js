export const formatDate = (timestamp) => {
  const date = new Date(timestamp);
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${hours}:${
    minutes < 10 ? '0' + minutes : minutes
  } | ${month}/${day}/${year}`;
};

export function sortUsers(users) {
  return users.sort((user1, user2) => {
    let user1Total = user1.questions.length + Object.keys(user1.answers).length;
    let user2Total = user2.questions.length + Object.keys(user2.answers).length;
    return user1Total < user2Total ? 1 : user1Total > user2Total ? -1 : 0;
  });
}

export function sortPolls(polls) {
  return polls.sort((poll1, poll2) => {
    return poll1.timestamp < poll2.timestamp
      ? 1
      : poll1.timestamp > poll2.timestamp
      ? -1
      : 0;
  });
}
