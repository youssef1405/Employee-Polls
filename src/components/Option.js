import styled from 'styled-components';
import { useDispatch, useSelector } from 'react-redux';
import { changePollStatus, addAnswer } from '../store';
import { VscCheck } from 'react-icons/vsc';

const Option = ({ text, option }) => {
  const dispatch = useDispatch();
  const question = useSelector((state) => state.polls.openedQuestion);
  const polls = useSelector((state) => state.polls.data);
  const { user } = useSelector((state) => state.users.currentUser);
  const { answers } = useSelector((state) => state.users.data[user]);
  const selected = answers[question.id] && option === answers[question.id];

  const handleClick = () => {
    dispatch(addAnswer({ qid: question.id, option }));
    dispatch(changePollStatus({ id: question.id, user, option }));
  };

  const answeredQuestions = polls.find((poll) => poll.id === question.id);
  const totalPollVotes =
    answeredQuestions['optionOne'].votes.length +
    answeredQuestions['optionTwo'].votes.length;

  return (
    <Wrapper>
      {selected && (
        <div className='stats'>
          <span>{answeredQuestions[option].votes.length} voted</span>
          <VscCheck className='selected' />
          <span>
            {(
              (answeredQuestions[option].votes.length / totalPollVotes) *
              100
            ).toFixed()}
            %
          </span>
        </div>
      )}

      <p>{text}</p>

      <button
        data-testid='vote-button'
        onClick={handleClick}
        disabled={answers[question.id]}
      >
        Vote
      </button>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  border: 1px solid #d6d6d6;
  width: 350px;
  text-align: center;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;

  .stats {
    display: flex;
    gap: 7rem;
    align-items: center;
  }

  p {
    font-size: 0.9rem;
    margin: 0;
    padding: 0.6rem 0;
  }
  .selected {
    color: #fff;
    background-color: #3452eb;
    border-radius: 50%;
    padding: 2px;
    margin-top: 5px;
  }

  button {
    width: 100%;
    border: none;
    padding: 0.7rem 0;
    background-color: #61a266;
    color: #fff;
    cursor: pointer;
    font-weight: 600;
  }
`;

export default Option;
