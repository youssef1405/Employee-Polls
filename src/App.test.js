import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { store } from './store';
import { _saveQuestion, _saveQuestionAnswer } from './_DATA';
import { Loading } from './components';
import { NewPoll, Login } from './pages';
import { MemoryRouter } from 'react-router-dom';
import { sortUsers, sortPolls } from './utils';

describe('_saveQuestion', () => {
  it('will return an object that has these properties: id, author, optionOne, optionTwo, timestamp', async () => {
    const poll = {
      author: 'Youssef',
      optionOneText: 'Java',
      optionTwoText: 'Python',
    };
    const response = await _saveQuestion(poll);
    expect(response.author).toEqual('Youssef');
    expect(response.optionOne).toEqual({ text: 'Java', votes: [] });
    expect(response.optionTwo).toEqual({ text: 'Python', votes: [] });
    expect(response.id).not.toBeNull();
    expect(response.timestamp).not.toBeNull();
  });

  it('will throw an error if one of the paramters is missing', async () => {
    const poll = { optionOneText: 'Java', optionTwoText: 'Python' };
    await expect(_saveQuestion(poll)).rejects.toEqual(
      'Please provide optionOneText, optionTwoText, and author'
    );
  });
});

describe('_saveQuestionAnswer', () => {
  it('will return true if proper object is passed to it', async () => {
    const response = await _saveQuestionAnswer({
      authedUser: 'zoshikanlu',
      qid: '8xf0y6ziyjabvozdd253nd',
      answer: 'optionOne',
    });
    expect(response).toBeTruthy();
  });

  it('will throw an error if one of the paramters of the object is missing', async () => {
    const answer = {
      qid: '8xf0y6ziyjabvozdd253nd',
      answer: 'optionOne',
    };
    await expect(_saveQuestionAnswer(answer)).rejects.toEqual(
      'Please provide authedUser, qid, and answer'
    );
  });
});

describe('Loading component', () => {
  it('matches the snapshot when text is passed', () => {
    const view = render(<Loading text='Loading' />);
    expect(view).toMatchSnapshot();
  });

  it('matches the snapshot when no props are passed', () => {
    const view = render(<Loading />);
    expect(view).toMatchSnapshot();
  });
});

describe('New Poll', () => {
  it('will show an error if option one and option two are not entered', () => {
    const view = render(
      <MemoryRouter>
        <Provider store={store}>
          <NewPoll />
        </Provider>
      </MemoryRouter>
    );
    const createPollButton = view.getByTestId('new-poll-btn');
    fireEvent.click(createPollButton);
    expect(view.getByTestId('error-msg')).toBeInTheDocument();
  });
});

describe('Login', () => {
  it('will  show error message if username or password is invalid', () => {
    const view = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );

    const loginButton = view.getByTestId('login-btn');
    const usernameInput = view.getByTestId('username-input');
    const passwordInput = view.getByTestId('password-input');

    fireEvent.change(usernameInput, { target: { value: 'username' } });
    fireEvent.change(passwordInput, { target: { value: 'password' } });
    fireEvent.click(loginButton);
    expect(view.queryByTestId('invalid')).toBeInTheDocument();
  });

  it('will show error message if username and password are blanks', () => {
    const view = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );

    const loginButton = view.getByTestId('login-btn');
    fireEvent.click(loginButton);
    expect(view.queryByTestId('invalid')).toBeInTheDocument();
  });
});

describe('sortUser', () => {
  it('will return the same size array and and user mtsamis will be at index 0', () => {
    const polls = [
      {
        id: '8xf0y6ziyjabvozdd253nd',
        author: 'sarahedo',
        timestamp: 1467166872634,
        optionOne: {
          votes: ['sarahedo'],
          text: 'Build our new application with Javascript',
        },
        optionTwo: {
          votes: [],
          text: 'Build our new application with Typescript',
        },
      },
      {
        id: '6ni6ok3ym7mf1p33lnez',
        author: 'mtsamis',
        timestamp: 1468479767190,
        optionOne: {
          votes: ['sarahedo'],
          text: 'hire more frontend developers',
        },
        optionTwo: {
          votes: ['mtsamis'],
          text: 'hire more backend developers',
        },
      },
    ];
    const result = sortPolls(polls);
    expect(result.length).toEqual(2);
    expect(result[0].author).toEqual('mtsamis');
  });
});
