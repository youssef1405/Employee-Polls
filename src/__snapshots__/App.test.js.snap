// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component matches the snapshot when no props are passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jSUZER jEoqXK"
      >
        <h2 />
        <svg
          class="spinner"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSUZER jEoqXK"
    >
      <h2 />
      <svg
        class="spinner"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Loading component matches the snapshot when text is passed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jSUZER jEoqXK"
      >
        <h2>
          Loading
        </h2>
        <svg
          class="spinner"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSUZER jEoqXK"
    >
      <h2>
        Loading
      </h2>
      <svg
        class="spinner"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
